File: crates/nil-core/src/village.rs
use crate::building::prelude::*;
use crate::player::PlayerId;
use crate::world::Coord;
use bon::Builder;

#[derive(Builder, Debug)]
pub struct Village {
  #[builder(start_fn, into)]
  pub coord: Coord,
  #[builder(into)]
  pub name: String,
  #[builder(into)]
  pub owner: Option<PlayerId>,
  #[builder(default)]
  pub infrastructure: Infrastructure,
}

#[derive(Debug, Default)]
pub struct Infrastructure {
  pub prefecture: Prefecture,
  pub academy: Academy,
  pub stable: Stable,
  pub sawmill: Sawmill,
  pub quarry: Quarry,
  pub iron_mine: IronMine,
  pub farm: Farm,
  pub warehouse: Warehouse,
  pub wall: Wall,
}

File: crates/nil/src/types/village.ts
import type { PlayerId } from './player';
import type * as building from './building';

export type Village = {
  readonly coord: Coord;
  readonly infrastructure: Infrastructure;
  readonly name: string;
  readonly owner: VillageOwner;
};

export type Infrastructure = {
  readonly academy: building.Academy;
  readonly farm: building.Farm;
  readonly ironMine: building.IronMine;
  readonly prefecture: building.Prefecture;
  readonly quarry: building.Quarry;
  readonly sawmill: building.Sawmill;
  readonly silo: building.Silo;
  readonly stable: building.Stable;
  readonly wall: building.Wall;
  readonly warehouse: building.Warehouse;
};

export type Coord = {
  readonly x: number;
  readonly y: number;
};

export type VillageOwner = VillageOwnerNone | VillageOwnerPlayer;

export type VillageOwnerNone = {
  readonly kind: 'none';
};

export type VillageOwnerPlayer = {
  readonly id: PlayerId;
  readonly kind: 'player';
};
